Feat: Upgrade Colony version via Multi-Sig

rumzledz opened the pull request:
## Description

This PR enables users to upgrade their Colony's version via a Multi-Sig motion.

![Colony Version MS](https://github.com/user-attachments/assets/c3330891-2918-43ca-90a0-d4e7c488c63f)

## Testing

As of the time this PR is created, the most recent Colony version is 16. In order to test a Colony version upgrade, we will need to deploy a newer Colony version since the Multi-Sig extension is only available from Colony version 16 upwards.

> [!IMPORTANT]
> Please test this with these block-ingestor updates: [block-ingestor#254  - Feat: Enable Colony version upgrade via a multi-sig motion](https://github.com/JoinColony/block-ingestor/pull/254)
> 1. Updates were made for `colony-js` so it's important you run `npm ci`
> 2. Run `npm run dev`
> 3. Wait for the auth proxy and block-ingestor services to be up and running
> 4. Open the `hardhat` console:
> ```bash
> npm run hardhat
> ```
> 5. Confirm that you are now in the `hardhat` console
> ![image](https://github.com/user-attachments/assets/9a108f0c-694f-4db4-a2ca-b4117286cd69)
> 6. Run the following commands to deploy Colony version 17:
> > _Deploying a Colony version will take a while. Feel free to brew a cup of coffee in the meantime... ☕_ 
> ```solidity
> deployNext = require("./scripts/deployOldUpgradeableVersion.js").deployNextColonyVersion
> cn = await artifacts.require("IColonyNetwork").at("0x777760996135F0791E2e1a74aFAa060711197777")
> await deployNext(cn)
> ```
> 7. For an extra peace of mind, you can check if you have successfully deployed Colony version 17:
> > _This should return a non-zero address_
> ```solidity
> await cn.getColonyVersionResolver(17)
> ```
> 8. Use a previous Colony version to set up your Colonies. In this PR, the previous Colony version is 16:
> ```console
> node scripts/create-data.js --usePreviousColonyVersion
> ```
> 9. Run the frontend
> ```console
> npm run frontend
> ```
> 10. Install & enable the Multi-SIg extension
> 11. Reload the page

> [!TIP]
> ## Troubleshooting
> ### OneTxPayment version error when finalising the motion
> If you see block ingestor throwing this error:
> ```console
> Error: Extension OneTxPayment in version 8 is not compatible with Colony version 16
> ```
> It might be that block ingestor is still running a stale version of colony-js. Delete your `node_modules` directory and reinstall all deps again:
> ```console
> rm -rf node_modules
> npm i
>```
>
> ### Errors when running the create-data script:
> ![Screenshot 2024-07-17 at 18 32 02](https://github.com/user-attachments/assets/9a236028-f2d6-44fd-9c12-017c5ba951f6)
> If you're having a bunch of these errors when running the create-data script and it crashes, stop all docker containers then open your Docker Dashboard and delete all of these in order
> 
> 1. Containers
> 2. Images
> 3. Volumes
> 4. Builds
>
> Quit and relaunch Docker then spin up your dev environment again:
> ```console
> npm run dev
> ```
> If you're still getting those errors, this will sound ridiculous but please restart your machine. It certainly worked for me at some point but I really have no clue about what's causing it 😅
>
> Please let me know if there's a better solution for this 🙏
>
> ### Multi-Sig extension installation hang
> On my machine, sometimes the installation just works and sometimes it hangs when you install it for the first time after spinning up your dev environment. If you're experiencing this, just reload the page and install it again.
>
> ### Invalid threshold on the Completed Action Component's Multi-Sig motion panel
> - Check your Colony's members list. If it's 0, chances are something in your dev environment broke 😞 Having no members means that we cannot verify if you have the required permissions to interact with the Multi-Sig panel. You'll have to restart everything again unfortunately.
> - If your members list is still in tact, give it 10 seconds then reload the page. The data should then come up.
>
> ### You get a blank page when visiting localhost:9091
> Re-bundle vite:
> ```console
> npm run frontend -- --force
> ```

> [!WARNING]
> I find that after a Colony has been updated to version 17, all other motions will stop working. Judging from the console errors, it seems that the colony-js lib needs to be fixed.

### Testing Steps

1. Bring up the Action Form and set it to "Upgrade Colony version"
2. Verify that
  - Current version is 16
  - New version is 17
3. Set the Decision method field to Multi-Sig
4. Fill in all other required fields
5. Submit the Form
6. On the CompletedAction component, verify that the Action Type is "Version Upgrade"
7. Finalise the motion
8. Verify that the message "Action was approved and executed." appears on the Completed Action component
9. Verify that the banner which prompts you to upgrade your Colony version isn't rendered anymore
10. Close the Action Form
11. Bring up the Action Form and set it to "Upgrade Colony version"
12. Verify that:
  - You see the message: "Your Colony version is up to date"
  - Current version is 17
  - New version is 17

13: Visit the GraphQL playground: http://localhost:20002
14. Run this GraphQL query:
```gql
query MyQuery {
  listColonies {
    items {
      version
      name
    }
  }
}
```
15. Verify that the Colony you upgraded has its version set to 17. If you updated Colony "planex", you should see:
```gql
{
  "data": {
    "listColonies": {
      "items": [
        {
          "version": 17,
          "name": "planex"
        },
        {
          "version": 16,
          "name": "wayne"
        }
      ]
    }
  }
}
```

## Diffs

**Changes** 🏗

The rootMotion saga thankfully already handles the multi-sig flow for the Colony Upgrade action so I just needed to make the necessary UI-related updates.

Resolves #2186

iamsamgibbs reviewed (changes requested):
Amazing job on this @rumzledz ! 🎉  I know it was a pain pulling this all together, but it's looking great.

I was able to deploy version 17 and use the create-data script to create colonies on version 16. I was able to install the multi-sig extension and create a version upgrade action using the multi-sig decision method.

<img width="1362" alt="Screenshot 2024-07-18 at 09 15 01" src="https://github.com/user-attachments/assets/2c8ccb72-7482-4428-8f0b-e58cf2e01154">

Approved and executed without issue.

<img width="1363" alt="Screenshot 2024-07-18 at 09 17 44" src="https://github.com/user-attachments/assets/618582df-7d22-454a-a6aa-5a8f59472144">

<img width="309" alt="Screenshot 2024-07-18 at 09 17 48" src="https://github.com/user-attachments/assets/65800c9e-e02f-4e33-a152-46e2a23a5fd8">

Activity list looks good:

<img width="744" alt="Screenshot 2024-07-18 at 09 15 59" src="https://github.com/user-attachments/assets/93c684b7-3c27-47da-a538-a9d6f8604af8">

The only minor pick up is the missing `CreatedIn` field from the completed action screen:

<img width="689" alt="Screenshot 2024-07-18 at 09 15 28" src="https://github.com/user-attachments/assets/ed4be7f3-2439-4da1-9746-ede70a83418b">

I also noticed this mistake I made with the create-data script change, could you fix it here? If `usePreviousColonyVersionArg` is true, then we shouldn't console log line 1488.

<img width="901" alt="Screenshot 2024-07-18 at 09 09 12" src="https://github.com/user-attachments/assets/b2810d12-f67c-4e08-a6b4-ed990f941501">

Is the `[colonyAddress]` -> `[AddressZero]` a breaking change? I think the other motions which use it will need updating too: editColonyMotion, managePermissionsMotion, manageReputationMotion, cancelMultiSig, finalizeMultiSig, setThreshold, voteOnMultiSig. (Manage Permissions will need a block-ingestor change `const isMultiSig = actionTarget !== colonyAddress;` to make this AddressZero too).

If it isn't a breaking change then we could leave this for the QoL issue.

The misalignment caused by the upgrade banner is fun too, I'll raise a separate issue for this:

<img width="1346" alt="Screenshot 2024-07-18 at 09 12 42" src="https://github.com/user-attachments/assets/32119a83-9d73-4366-80c5-e654cbf5050d">

rumzledz commented:
Thanks for reviewing this @iamsamgibbs! I'm just gonna re-post the message here for others who are curious as well:

The colonyAddress -> addressZero change (which is not viewable on my PR anymore since the `feat/multi-sig` branch has these updates already) is a required change because on the contract side, it now goes:

If you're making a motion targeting the colony, use address zero.

Which should now be in line with how reputation motions work.

I've updated the Completed Action for to show the CreatedIn field:

![Screenshot 2024-07-18 at 11 19 13](https://github.com/user-attachments/assets/d9754c97-3bf8-41b1-8b1d-33e55c958979)

And updated the create-data script copy:

![Screenshot 2024-07-18 at 11 19 33](https://github.com/user-attachments/assets/cfb2f2bb-5544-4eb2-8c81-76ed3508acbf)

bassgeta reviewed (approved):
Thank you for the detailed testing steps, got the banner to show up :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/75aacd23-8cf9-44dd-9804-275083bf5030)

The header is a bit hilarious, but it's not your fault :D
![image](https://github.com/user-attachments/assets/1e319fc4-2aa5-47d8-bb8b-f9c6c8db6196)
![image](https://github.com/user-attachments/assets/86455e87-098c-4cef-9da3-c6293f1dbd67)

Nice going, let's merge this!

davecreaser reviewed (approved):
Great work @rumzledz! I bet it was a lot of fun when you first picked this up trying to figure out how to even start developing this 🤣 Really nice job. Code is all good, and it works great when testing (barring the banner misalignment that Sam has already pointed out, but that's not part of this PR's remit).

Here's my testing, showing that it worked flawlessly:
<img width="1920" alt="Screenshot 2024-07-22 at 11 35 09" src="https://github.com/user-attachments/assets/839d1660-d36e-4447-81ec-5ffc3a140b40">
<img width="1920" alt="Screenshot 2024-07-22 at 11 37 33" src="https://github.com/user-attachments/assets/b343ea6b-8daa-4214-9277-71f8cb6dcd76">
<img width="1920" alt="Screenshot 2024-07-22 at 11 38 04" src="https://github.com/user-attachments/assets/22fa076d-b292-4a8b-998d-305d2737b41d">
<img width="1920" alt="Screenshot 2024-07-22 at 11 38 32" src="https://github.com/user-attachments/assets/03ded8a8-bbe7-41ef-b71f-fb8589459fe1">
<img width="1920" alt="Screenshot 2024-07-22 at 11 38 55" src="https://github.com/user-attachments/assets/e14e7eec-654c-436c-a8dd-bb9e989fca66">
<img width="1920" alt="Screenshot 2024-07-22 at 11 39 13" src="https://github.com/user-attachments/assets/166c873c-8f15-412d-a6a2-8f4f70b3d667">

rumzledz commented:
> I bet it was a lot of fun when you first picked this up trying to figure out how to even start developing this

It was quite the adventure! These two emojis are pretty much summing it up 😭 😂

iamsamgibbs reviewed (approved):
Nice job! My previous issues have been addressed, get it merged!

<img width="1318" alt="Screenshot 2024-07-23 at 09 16 49" src="https://github.com/user-attachments/assets/9e3dba0b-c220-4ae4-b219-f77599e51f8f">

<img width="748" alt="Screenshot 2024-07-23 at 09 15 59" src="https://github.com/user-attachments/assets/9320051b-3fdf-4cff-9ffa-d222d8aeafea">

<img width="1055" alt="Screenshot 2024-07-23 at 09 15 54" src="https://github.com/user-attachments/assets/a7d1ebc5-a47e-4e55-896c-7102dd539ad7">
